Merge feature-advanced_activity to dev

dmitry-zhbankov opened the pull request.

dmitry-zhbankov committed:
Add code

dmitry-zhbankov committed:
Add wcf service

dmitry-zhbankov committed:
Add HttpClient request

dmitry-zhbankov committed:
Add request to GitLab

dmitry-zhbankov committed:
Retrieve commits from GitLab api

dmitry-zhbankov committed:
Get repositories from config

dmitry-zhbankov committed:
Complete dev activities

dmitry-zhbankov committed:
Refactor methods
Add subsequent requests if server doesn't respond

dmitry-zhbankov committed:
Fix commits api url

dmitry-zhbankov committed:
Get mentoring activities

dmitry-zhbankov committed:
Get Activities SP List from service

dmitry-zhbankov committed:
Update Activities SP List

dmitry-zhbankov committed:
Add logger

dmitry-zhbankov committed:
Add error handling
Improve logging

dmitry-zhbankov committed:
Improve error handling

dmitry-zhbankov committed:
Refactor code

dmitry-zhbankov committed:
Make renaming
Add user lookup

dmitry-zhbankov committed:
Add base ActivitySource class

dmitry-zhbankov committed:
Add ensure dictionary

dmitry-zhbankov committed:
Add path field to ActivityInfo

dmitry-zhbankov committed:
Add SPHelper

dmitry-zhbankov committed:
Add API helper
GitLabActivitySource as partial class

dmitry-zhbankov committed:
FIx GetSpMembers method

dmitry-zhbankov committed:
Add activity user check

dmitry-zhbankov committed:
Set lookup user value

dmitry-zhbankov committed:
Fix duplicate keys issue

dmitry-zhbankov committed:
Add activity checks

dmitry-zhbankov committed:
Extract EnsureService class

dmitry-zhbankov committed:
Add EnsureService tests

dmitry-zhbankov committed:
Add Ensure() method test.

dmitry-zhbankov committed:
Adjust timer

valerydolyuk commented on Test.Activities.Automation.TimerJob/Features/TimerJobFeature/TimerJobFeature.EventReceiver.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            var webApp = properties.Feature.Parent as SPWebApplication;
+
+            foreach (var job in webApp.JobDefinitions)
```
These lines might be moved in a separate method. There is duplicate below

valerydolyuk commented on Test.Activities.Automation.TimerJob/Features/TimerJobFeature/TimerJobFeature.EventReceiver.cs:
```diff
@@ -0,0 +1,74 @@
+    [Guid("353e1448-874f-405e-af35-3a78c37e98af")]
+    public class TimerJobFeatureEventReceiver : SPFeatureReceiver
+    {
+        const string docJobName = "Test.Activities.Automation.Job";
```
1. Write constants in a PascalCase
2. It seems like you copied variable name `docJobName` from some where. You should use other name here

valerydolyuk commented on Test.Activities.Automation.TimerJob/Features/TimerJobFeature/TimerJobFeature.EventReceiver.cs:
```diff
@@ -0,0 +1,74 @@
+            var docJob = new TimerJob(docJobName, webApp,
+                SPServer.Local, SPJobLockType.Job);
+
+            var schedule = new SPDailySchedule();
```
Note: you may set some schedule props (hours, minutes to start, etc.)

valerydolyuk commented on Test.Activities.Automation.TimerJob/TimerJob.cs:
```diff
@@ -0,0 +1,89 @@
+
+namespace Test.Activities.Automation.TimerJob
+{
+    public class TimerJob : SPJobDefinition
```
Other than just `TimerJob` name must be used for class and file names

valerydolyuk commented on Test.Activities.Automation.TimerJob/TimerJob.cs:
```diff
@@ -0,0 +1,89 @@
+
+                for (var i = 0; i < RequestAttempts; i++)
+                {
+                    if (await APIHelper.PostJsonAsync(uri, str) == HttpStatusCode.OK) return;
```
In case of error you lose original exception. Always log it.

dmitry-zhbankov replied:
In case of exception it bubbles up to

dmitry-zhbankov replied:
In case of exception it bubbles up from PostJsonAsync() and is handled by SendActivities() in the catch block. The exception handling will not work properly when using async method without "await" keyword or void async method.

valerydolyuk replied:
I mean if status code is not 200. For an instance, if service URL is not valid, etc.

valerydolyuk commented on Test.Activities.Automation.WCFService/ISAPI/AutomationService.svc.cs:
```diff
@@ -0,0 +1,58 @@
+            }
+        }
+
+        public HttpStatusCode FillActivities(IEnumerable<ActivityInfo> activities)
```
HTTP/s request has separate code value in the header, you don't need to send it in the body of the response.
In order to set status code you should use:
`
WebOperationContext ctx = WebOperationContext.Current;
ctx.OutgoingResponse.StatusCode = System.Net.HttpStatusCode.OK;
`

dmitry-zhbankov replied:
I agree with you. Earlier I returned true/false from this method. But then I decided to change true/false to status code and forgot to put it in the head instead of the body.

valerydolyuk commented on Test.Activities.Automation.WCFService/ISAPI/AutomationService.svc.cs:
```diff
@@ -0,0 +1,58 @@
+            }
+            catch
+            {
+                statusCode = HttpStatusCode.InternalServerError;
```
Never return 5XX+ status code in the service methods. For errors in code 4XX should be used.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Classes/ActivityInfo.cs:
```diff
@@ -0,0 +1,42 @@
+    [DataContract]
+    public class ActivityInfo
+    {
+        public int? UserId { get; set; }
```
You must create 2 separate classes: one with `UserId` which is used by Fetch/Sync service and the second one with `UserEmail` which is used as a parameter in the WCF service.
Because you combined them you had to
1. create `CheckActivityUser` method in the Ensure service and modify parameter inside it;
2. make `UserId` as nullable;
3. define `MentorEmail` and  `RootMentorEmail` properties in `Member` class;
4. may be smth else but 3 disadvantages are enough to refactor this.

dmitry-zhbankov replied:
Both WCF service and Sync service use ActivityInfo with UserId. So I can't split ActivityInfo class in that case. The only case to use ActivityInfo with UserEmail is in the GetDevActivities(). Hence, I  am going to use separate class for that case.
The CheckActivityUser() method is used not only for getting UserId by email but also for validating existing users and etc. So I cant remove that method completely.

valerydolyuk replied:
You are completely right: you have to create new class for WCF Service with Email.
As for removing `CheckActivityUser` - you may find more details in [this comment](https://github.com/dmitry-zhbankov/Test.Activities.Automation/pull/2/files#r381602605) (SyncActivityService.cs file, line 46)

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Classes/Member.cs:
```diff
@@ -0,0 +1,21 @@
+
+        public int? MentorLookupId { get; set; }
+
+        public string MentorEmail { get; set; }
```
Avoid cases where you need `Email`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Classes/SpActivity.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        public int Id { get; set; }
+
+        public int? MentorId { get; set; }
```
You may use `Member` class instead of duplicate property definitions

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Classes/SpActivity.cs:
```diff
@@ -0,0 +1,50 @@
+            if (obj is SpActivity otherObj)
+            {
+                return MentorId == otherObj.MentorId &&
+                       MentorLookupId == otherObj.MentorLookupId &&
```
Use just tab indentation instead of 7 spaces.

dmitry-zhbankov replied:
Ok, as you want. This is just a Resharper suggestion. Also, I consider the rule "Line up parameters to a method that are on new lines".

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.ActivityKey.cs:
```diff
@@ -0,0 +1,29 @@
+
+            public override int GetHashCode()
+            {
+                return UserId ^ Year ^ Month;
```
This is not unique value: `6 ^ 2020 ^ 1 == 2019 == 5 ^ 2020 ^ 2`.

dmitry-zhbankov replied:
In that case the Equals() method performs strict comparison

valerydolyuk replied:
I understand why you need to override this method, but current implementation doesn't return a unique value. I wrote an example where you get the same result for two different parameters. Implementation must be changed.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.ActivityValue.cs:
```diff
@@ -0,0 +1,26 @@
+        {
+            public int ActivityId { get; set; }
+
+            public int? MentorId { get; set; }
```
You may reuse `Member` class here.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Sources/ActivitySource.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+        }
+
+        public abstract IEnumerable<ActivityInfo> FetchActivity();
```
`FetchActivities` - plural

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Sources/GitLabActivitySource/GitLabActivitySource.cs:
```diff
@@ -0,0 +1,169 @@
+                using (var web = site.OpenWeb(Constants.Web))
+                {
+                    var configList = web.Lists[Constants.Lists.Configurations];
+                    configItems = configList.GetItems().Cast<SPListItem>();
```
1. Use `SPQuery` here
2. Move `GetConfig` method in a separate utility/service

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Sources/GitLabActivitySource/GitLabActivitySource.cs:
```diff
@@ -0,0 +1,169 @@
+            }
+            catch (Exception e)
+            {
+                _logger.LogError($"Fetching GitLab activity failed. {e.Message}");
```
Always log all error content including `StackTrace`, i.e. just `{e}` instead of `e.Message`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Sources/GitLabActivitySource/GitLabActivitySource.cs:
```diff
@@ -0,0 +1,169 @@
+                            {
+                                var commits = GetBranchCommits(client, branch, repo);
+
+                                branch.Commits = new List<Commit>(commits);
```
Calling just `ToList` method would be more readable

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Sources/SPCalendarActivitySource.cs:
```diff
@@ -0,0 +1,93 @@
+                    _mentorList = web.Lists[Constants.Lists.Mentors];
+                }
+
+                var eventsCollection = mentoringCalendarList.GetItems();
```
Use `SPQuery`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Helpers/APIHelper.cs:
```diff
@@ -0,0 +1,77 @@
+            var response = client.GetAsync(url).Result;
+            var stream = response.Content.ReadAsStreamAsync().Result;
+
+            var res = new List<T>();
```
It would be more readable if you move this initialization to top

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Helpers/SPHelper.cs:
```diff
@@ -0,0 +1,95 @@
+            var rootMentorField = originList.Fields.GetField(originField);
+            var rootMentorItem = originList.GetItemById(userFieldLookUpValue.LookupId);
+
+            var rootMentorFieldValue =
```
1. You are already have `GetUserValue` utility
2. `rootMentor` name is not correct

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Helpers/SPHelper.cs:
```diff
@@ -0,0 +1,95 @@
+                    SPFieldMultiChoiceValue;
+
+            var res = new List<string>();
+            for (int i = 0; i < fieldValue.Count; i++)
```
Why don't just use `Cast<string>()`?

dmitry-zhbankov replied:
SPFieldMultiChoiceValue is not IEnumerable

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Logger/ULSLogger.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Test.Activities.Automation.ActivityLib.Models
+{
+    public class ULSLogger : ILogger
```
There are a lot of classes which are not models. They are services, helpers, utils. So they should be moved in a separate folder

valerydolyuk commented on Test.Activities.Automation.Test/Program.cs:
```diff
@@ -0,0 +1,46 @@
+{
+    class Program
+    {
+        static void Main(string[] args)
```
In order to do testing you must know what should be logged into a console and you have to manually check result. In your implementation it is not an automated testing this is anyway manual tests.
If you don't use any testing framework, your code nust be (just an example):
1. define int prop `ErrorsCount` in `DebugLogger`
2. increase value of this prop inside `LogError` method
3. in `Main` check value of `ErrorsCount`. If it equals zero - print "success", otherwise "error"

valerydolyuk commented on Test.Activities.Automation.ActivityLib.Tests/Models/Activity/Services/EnsureService/EnsureServiceTests.cs:
```diff
@@ -0,0 +1,411 @@
+            var maxDate = DateTime.Now.Date;
+            var minDate = DateTime.Now.AddMonths(-1).Date;
+
+            IEnumerable<ActivityInfo> actualActivities = new List<ActivityInfo>()
```
Here and below you may use some `GetFakeActivitiesData`, `GetFakeMembersData` methods. This decreases size of test methods and improves readability.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/ActivityService.cs:
```diff
@@ -0,0 +1,361 @@
+
+namespace Test.Activities.Automation.ActivityLib.Models
+{
+    public class ActivityService
```
Is this class used?

dmitry-zhbankov replied:
Oh, I've excluded this class from my project but forgot to remove it from Git.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/FetchActivityService.cs:
```diff
@@ -0,0 +1,61 @@
+
+            var mentoringActivities = GetMentoringActivities();
+
+            return devActivities.Concat(mentoringActivities).ToList();
```
You implement inheritance but do not use inheritance features. Very simplified example:

`public IEnumerable<ActivityInfo> FetchActivities()`
`{`
`     _logger?.LogInformation("Fetching activities");`
`    return (new List<ActivitySource>{ new GitLabActivitySource(_logger), new SPCalendarActivitySource(_logger)}).SelectAll(source => source.FetchActivity());`
`}`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+            _logger = logger;
+        }
+
+        public void CheckActivityDate(ref IEnumerable<ActivityInfo> activities, DateTime minDate, DateTime maxDate)
```
Why do you always use `ref`?

valerydolyuk replied:
Oh... it seems like I understand! You must avoid use IEnumerble to often. You must call ToList/ToArray sometimes.
[article sample to read](https://habr.com/ru/post/193774/)

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+
+                if (member == null) continue;
+
+                activity.Paths = new List<string>(member.Paths);
```
Do not modify parameter inside a method.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+                    continue;
+                }
+
+                activity.UserId = id;
```
Do not modify parameter inside a method.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+        private int? GetUserIdByEmail(string email, IEnumerable<Member> members)
```
Remove usage of email in Sync part of the project

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+
+                var ensureService = new EnsureService(_logger);
+
+                ensureService.CheckActivityDate(ref activities, minDate, maxDate);
```
You implemented complex dictionary key in `EnsureService` which depends on month and year. But here you remove new activities which are for other monts. I'm not sure, but it seems like if you remove this line and `CheckActivityDate` method, code still works. Additionaly: `var minDate = activities.Min(x => x.Date);` and `var maxDate = activities.Max(x => x.Date);`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+                ensureService.CheckActivityDate(ref activities, minDate, maxDate);
+
+                using (var site = new SPSite(Constants.Host))
+                using (var web = site.OpenWeb(Constants.Web))
```
You may initialize `SPWeb` only once: in the TimerJob. Sources, Services may take it as a constructors' parameter. In WCF Service you have `SPContext.Current.Web` (note: you do not need to dispose `SPContext` properties)

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+
+                    var members = GetSpMembers(web);
+
+                    ensureService.CheckActivityUser(ref activities, members);
```
Remove usage of `CheckActivityUser`.
Replace with smth like:
`var memberUserIds = members.SelectAll(m => new[] {m.MentorId, m.RootMentorId}).Where(id => id.HasValue).Select(id => id.Value).Distinct().ToList()`
`activities = activities.Where(a => memberUserIds.Contains(a.UserId))`
BTW: `MentorId` and `RootMentorId` should be replaced by `UserId`

valerydolyuk replied:
After removing these `Id` props and avoiding usage of nullable types code will look good:
`var memberUserIds = members.Select(m => m.UserId).Distinct().ToList();`
`activities = activities.Where(a => memberUserIds.Contains(a.UserId))`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+
+                    ensureService.CheckActivityUser(ref activities, members);
+
+                    ensureService.CheckActivityPaths(ref activities, members);
```
Remove usage of `CheckActivityPaths`.
I don't see great way to do this right now. Temperary, thiss logic might be moved inside `Ensure` method at line 115 where you have both: `activities` and `members`.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Classes/Member.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class Member
+    {
+        public int? MentorId { get; set; }
```
Instead of nullable `MentorId` and `RootMentorId` properties, just define one nonnullable `UserId`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+
+                    var paths = SPHelper.GetMultiChoiceValue(item, Constants.Activity.Paths);
+
+                    members.Add(new Member
```
You may specify `Member Member.FromMentor(SPListItem item)` and `Member Member.FromRootMentor(SPListItem item)` static methods and move common code to a separate method.

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+                var spListActivities = web.Lists.TryGetList(Constants.Lists.Activities);
+                if (spListActivities == null) throw new Exception("Getting SP activity list failed");
+
+                var spListMentors = web.Lists.TryGetList(Constants.Lists.Mentors);
```
You get all members twice: here and in `GetMembers` method. Let's add one parameter `List<Member> members` to this method

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+                    var activities = SPHelper.GetMultiChoiceValue(item, Constants.Activity.Activities);
+                    var paths = SPHelper.GetMultiChoiceValue(item, Constants.Activity.Paths);
+
+                    spActivities.Add(new SpActivity
```
You may move this logic into a constructor of `SpActivity`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+                }
+
+                var itemsToUpdate = dict.Where(x => x.Value.IsModified)
+                    .Select(x => new SpActivity()
```
You may move this logic into a constructor of `SpActivity`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+            {
+                _logger?.LogInformation("Updating SP activities list");
+
+                var spList = web.Lists.TryGetList(Constants.Lists.Activities);
```
This is the second place where you get this list in this class

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/SyncActivityService.cs:
```diff
@@ -0,0 +1,246 @@
+
+                foreach (var item in itemsToUpdate)
+                {
+                    if (item.IsNew)
```
You may move this logic and `InsertActivity` and `UpdateActivity` methods to new method of `SpActivity` class

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+            {
+                _logger?.LogInformation("Ensuring activities");
+
+                var dict = InitDictionary(spActivities);
```
Actually, all code which works with SpActivity might be moved into `EnsureService`. Here you may retrieve all existing activities. Inside `SyncActivityService` you may just call `Save` method of `EnsureService`

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+                    Month = spActivity.Month
+                };
+
+                var value = new ActivityValue()
```
Why don't use constructor?

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+                        break;
+                }
+
+                var newValue = new ActivityValue()
```
Why don't use constructor?

valerydolyuk commented on Test.Activities.Automation.ActivityLib/Models/Activity/Services/EnsureService/EnsureService.cs:
```diff
@@ -0,0 +1,252 @@
+
+                var newKey = key;
+
+                switch (activity.Activity)
```
I don't think that this service should know this business logic. This service is not responsible for determination of activity type. Just remove this validation.

valerydolyuk reviewed (changes requested):
Please, resolve issues marked with "thumb up" icon. Skip others.

dmitry-zhbankov committed:
Add preliminary changes:
* Move SP methods to corresponding classes
* Refactor code
* Make renaming
* Replace getting activities methods with base ActivitySource method
* Reuse Member class
* Add ActivityInfoEmail class

dmitry-zhbankov committed:
Replace lists usage with spMembers

dmitry-zhbankov committed:
Add SP Queries
Remove unused methods

dmitry-zhbankov committed:
Fix paths serialization
Edit ServiceUrl

dmitry-zhbankov committed:
Pass SPWeb instead of SPList to methods
Restore web's AllowUnsafeUpdates property after setting
Rename "Sp" to "SP"
Remove static loggers
